src/librustc_typeck/check/writeback.rs

```diff
@@ -190,9 +190,23 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
         if let hir::ExprKind::Index(ref base, ref index) = e.node {
             let mut tables = self.fcx.tables.borrow_mut();
 
-            // All valid indexing looks like this; might encounter non-valid indexes at this point
-            if let ty::Ref(_, base_ty, _) = tables.expr_ty_adjusted(&base).sty {
-                let index_ty = tables.expr_ty_adjusted(&index);
+            // All valid indexing looks like this; might encounter non-valid indexes at this point.
+            let base_ty = tables.expr_ty_adjusted_opt(&base).map(|t| &t.sty);
+            if base_ty.is_none() {
+                self.tcx().sess.delay_span_bug(e.span, &format!("bad base: `{:?}`", base));
+                return;
+            }
+            if let Some(ty::Ref(_, base_ty, _)) = base_ty {
```

Comment:
We are pattern matching on `Some(...)` here anyways so the early return and checking `.is_none()` feels odd -- just use `else {` instead and you won't need `return;` at all.

Reply:
I originally didn't have the `return` because it _is_ unnecessary, but then I thought I'd make it clear that we should _not_ continue processing in that case in case more logic is added in the currently inexistent else case below.

Reply:
In case you want to have the `None` branch up-front, use a `match` and put `None =>` first then.